src/Filter.ts

```diff
@@ -0,0 +1,110 @@
+export interface FilterArgs {
+	[ name: string ]: Filter | RegExp | string | (() => boolean);
+}
+
+export default class Filter {
+	[index: string]: any;
+	type: string;
+	args: any[];
+
+	constructor();
+	constructor(filterArg: string);
+	constructor(filterArg: () => any);
+	constructor(filterArg: FilterArgs);
+	constructor(filterArg?: any) {
+		if (typeof filterArg === 'string' || typeof filterArg === 'function') {
+			this.type = typeof filterArg;
+			this.args = [filterArg];
+		}
+		else if (filterArg) {
+			var filter = this;
+			// construct a filter based on the query object
+			for (var key in filterArg) {
+				var value = filterArg[key];
+				if (value instanceof Filter) {
+					// fully construct the filter from the single arg
+					filter = filter[value.type](key, value.args[0]);
+				}
+				else if ((<any>value) && typeof (<any>value).test === 'function') {
+					// support regex and functions
+					filter = filter.match(key, value);
+				}
+				else {
+					filter = filter.eq(key, value);
+				}
+			}
+			this.type = filter.type;
+			this.args = filter.args;
+		}
+	}
+
+	protected _createFilter(key: any, target: any, type: string) {
+		var filter = new Filter();
+		filter.type = type;
+		filter.args = [key, target];
+		if (this instanceof Filter && this.type) {
+			// we are chaining, so combine with an and operator
+			return this.and(this, filter);
+		}
+		return filter;
+	}
+
+	protected _createLogicalFilter(type: string, args: Array<Filter | string>): Filter {
+		let argsArray = args.map((arg) => {
+			return arg instanceof Filter ? arg : new Filter(<string>arg);
```

Comment:
Is this cast needed `new Filter(<string>arg)`? Given the definition above I think you can get away without it.